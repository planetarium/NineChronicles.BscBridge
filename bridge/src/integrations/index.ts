export interface Integration {
    error(summary: string, error: Record<string, any>): void;
}
