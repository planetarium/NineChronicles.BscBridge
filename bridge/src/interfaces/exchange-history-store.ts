import { TransactionStatus } from "../types/transaction-status";

export interface ExchangeHistory {
  network: string;
  tx_id: string;
  sender: string;
  recipient: string;
  timestamp: string;
  amount: number;
  status: TransactionStatus;
}

export interface IExchangeHistoryStore {
  put(history: ExchangeHistory): Promise<void>;
  exist(tx_id: string): Promise<boolean>;
  updateStatus(
    tx_id: string,
    status: TransactionStatus.COMPLETED | TransactionStatus.FAILED
  ): Promise<void>;

  transferredAmountInLast24Hours(
    network: string,
    sender: string
  ): Promise<number>;

  getPendingTransactions(): Promise<ExchangeHistory[]>;
}
