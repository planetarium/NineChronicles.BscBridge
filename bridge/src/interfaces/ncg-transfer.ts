import { TxId } from "../types/txid";

export interface INCGTransfer {
    transfer(
        address: string,
        amount: string,
        memo: string | null
    ): Promise<TxId>;
}
