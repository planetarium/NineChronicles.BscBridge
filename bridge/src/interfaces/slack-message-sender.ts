import { ChatPostMessageResponse } from "@slack/web-api";
import { Message } from "../messages";

export interface ISlackMessageSender {
  sendMessage<T extends Message>(message: T): Promise<ChatPostMessageResponse>;
}
