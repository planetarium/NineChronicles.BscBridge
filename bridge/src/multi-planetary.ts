export interface IPlanetIds {
  odin: string;
  heimdall: string;
}

export interface IPlanetVaultAddress {
  heimdall: string;
}

const MAIN_PLANET_NAME = "odin";
type PlanetName = keyof IPlanetIds;

export class MultiPlanetary {
  private readonly _planetIds: IPlanetIds;
  private readonly _planetVaultAddresses: IPlanetVaultAddress;

  constructor(
    planetIds: IPlanetIds,
    planetVaultAddresses: IPlanetVaultAddress
  ) {
    this._planetIds = planetIds;
    this._planetVaultAddresses = planetVaultAddresses;
  }

  isMultiPlanetRequestType(_to: string): boolean {
    const multiPlanetIdRegex = /^[0-9]0{10}[0-9]$/;
    return multiPlanetIdRegex.test(_to.substring(2, 14));
  }

  getRequestPlanetName(_to: string): PlanetName {
    let planetName = MAIN_PLANET_NAME;

    if (!this.isMultiPlanetRequestType(_to)) {
      return MAIN_PLANET_NAME;
    }

    /**
     * If requestPlanetId is not in this._planetIds ( Invalid Multi-Planet Id )
     * Then send request to "odin".
     */
    for (const key of Object.keys(this._planetIds)) {
      if (_to.startsWith(this._planetIds[key as keyof IPlanetIds])) {
        planetName = key;
        break;
      }
    }

    return planetName as PlanetName;
  }

  isMainPlanetRequest(planetName: string): boolean {
    return planetName === MAIN_PLANET_NAME;
  }

  getPlanetVaultAddress(planetName: string): string {
    return this._planetVaultAddresses[planetName as keyof IPlanetVaultAddress];
  }
}
