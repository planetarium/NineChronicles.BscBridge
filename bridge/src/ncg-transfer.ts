import { IHeadlessGraphQLClient } from "./interfaces/headless-graphql-client";
import { INCGTransfer } from "./interfaces/ncg-transfer";
import { TxId } from "./types/txid";

export class NCGTransfer implements INCGTransfer {
    private readonly _headlessGraphQLCLient: IHeadlessGraphQLClient;

    /**
     * The address of sender;
     */
    private readonly _address: string;

    constructor(
        headlessGraphQLCLient: IHeadlessGraphQLClient,
        address: string
    ) {
        this._headlessGraphQLCLient = headlessGraphQLCLient;
        this._address = address;
    }

    async transfer(
        address: string,
        amount: string,
        memo: string | null
    ): Promise<TxId> {
        const nextTxNonce = await this._headlessGraphQLCLient.getNextTxNonce(
            this._address
        );
        return await this._headlessGraphQLCLient.transfer(
            address,
            amount,
            nextTxNonce,
            memo
        );
    }
}
