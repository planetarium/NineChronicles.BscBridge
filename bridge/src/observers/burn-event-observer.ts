import { IObserver } from ".";
import { BlockHash } from "../types/block-hash";
import { EventData } from "web3-eth-contract";
import { TransactionLocation } from "../types/transaction-location";
import { BurnEventResult } from "../types/burn-event-result";
import { INCGTransfer } from "../interfaces/ncg-transfer";
import { OpenSearchClient } from "../opensearch-client";
import { IMonitorStateStore } from "../interfaces/monitor-state-store";
import { UnwrappedEvent } from "../messages/unwrapped-event";
import Decimal from "decimal.js";
import { UnwrappingFailureEvent } from "../messages/unwrapping-failure-event";
import { Integration } from "../integrations";
import { ISlackMessageSender } from "../interfaces/slack-message-sender";
import { IExchangeHistoryStore } from "../interfaces/exchange-history-store";
import { UnwrappingRetryIgnoreEvent } from "../messages/unwrapping-retry-ignore-event";
import { SpreadsheetClient } from "../spreadsheet-client";
import { MultiPlanetary } from "../multi-planetary";
import { TransactionStatus } from "../types/transaction-status";

export class BscBurnEventObserver
  implements
    IObserver<{
      blockHash: BlockHash;
      events: (EventData & TransactionLocation)[];
    }>
{
  private readonly _ncgTransfer: INCGTransfer;
  private readonly _opensearchClient: OpenSearchClient;
  private readonly _spreadsheetClient: SpreadsheetClient;
  private readonly _slackMessageSender: ISlackMessageSender;
  private readonly _monitorStateStore: IMonitorStateStore;
  private readonly _exchangeHistoryStore: IExchangeHistoryStore;
  private readonly _explorerUrl: string;
  private readonly _ncscanUrl: string | undefined;
  private readonly _useNcscan: boolean;
  private readonly _etherscanUrl: string;
  private readonly _integration: Integration;
  private readonly _multiPlanetary: MultiPlanetary;
  private readonly _failureSubscribers: string;
  private readonly _opensearchMigrationClient: OpenSearchClient;

  constructor(
    ncgTransfer: INCGTransfer,
    slackMessageSender: ISlackMessageSender,
    opensearchClient: OpenSearchClient,
    spreadsheetClient: SpreadsheetClient,
    monitorStateStore: IMonitorStateStore,
    exchangeHistoryStore: IExchangeHistoryStore,
    explorerUrl: string,
    ncscanUrl: string | undefined,
    useNcscan: boolean,
    etherscanUrl: string,
    integration: Integration,
    multiPlanetary: MultiPlanetary,
    failureSubscribers: string,
    opensearchMigrationClient: OpenSearchClient
  ) {
    this._ncgTransfer = ncgTransfer;
    this._slackMessageSender = slackMessageSender;
    this._opensearchClient = opensearchClient;
    this._spreadsheetClient = spreadsheetClient;
    this._monitorStateStore = monitorStateStore;
    this._exchangeHistoryStore = exchangeHistoryStore;
    this._explorerUrl = explorerUrl;
    this._ncscanUrl = ncscanUrl;
    this._useNcscan = useNcscan;
    this._etherscanUrl = etherscanUrl;
    this._integration = integration;
    this._multiPlanetary = multiPlanetary;
    this._failureSubscribers = failureSubscribers;
    this._opensearchMigrationClient = opensearchMigrationClient;
  }

  async notify(data: {
    blockHash: BlockHash;
    events: (EventData & TransactionLocation)[];
  }): Promise<void> {
    const { blockHash, events } = data;
    if (events.length === 0) {
      await this._monitorStateStore.store("bsc", {
        blockHash,
        txId: null,
      });
    }

    for (const { returnValues, transactionHash, blockHash } of events) {
      const {
        _sender: sender,
        _to,
        amount: burnedWrappedNcgAmountString,
      } = returnValues as BurnEventResult;

      // Added logging for checking _to arg
      console.log("_to", _to);

      const isMultiPlanetRequestType =
        this._multiPlanetary.isMultiPlanetRequestType(_to);

      const requestPlanetName = this._multiPlanetary.getRequestPlanetName(_to);

      const user9cAddress = isMultiPlanetRequestType
        ? "0x" + _to.substring(14, 54)
        : _to.substring(0, 42);

      const amount = new Decimal(burnedWrappedNcgAmountString).div(
        new Decimal(10).pow(18)
      );
      const amountString = amount.toFixed(2, Decimal.ROUND_DOWN);

      if (await this._exchangeHistoryStore.exist(transactionHash)) {
        this._slackMessageSender.sendMessage(
          new UnwrappingRetryIgnoreEvent(transactionHash)
        );
        this._opensearchClient.to_opensearch("error", {
          content: "wNCG -> NCG request failure",
          cause: "Exchange history exist",
          ethereumTxId: transactionHash,
          sender: sender,
          recipient: user9cAddress,
          amount: amountString,
          network: "BSC",
        });
        this._opensearchMigrationClient.to_opensearch("error", {
          content: "wNCG -> NCG request failure",
          cause: "Exchange history exist",
          ethereumTxId: transactionHash,
          sender: sender,
          recipient: user9cAddress,
          amount: amountString,
          network: "BSC",
        });
        continue;
      }

      await this._exchangeHistoryStore.put({
        network: "bsc",
        tx_id: transactionHash,
        sender,
        recipient: user9cAddress,
        timestamp: new Date().toISOString(),
        amount: parseFloat(amountString),
        status: TransactionStatus.PENDING,
      });

      try {
        console.log("Process Ethereum transaction", transactionHash);
        const isMainPlanetRequest =
          this._multiPlanetary.isMainPlanetRequest(requestPlanetName);
        if (!isMainPlanetRequest) {
          console.log(`Send to other planet - ${requestPlanetName}`);
        }
        /**
         * If User send wNCG to other planet
         * recipient is other planet's vault address.
         * memo is user's other planet's 9c Address.
         */
        const recipient =
          isMultiPlanetRequestType && !isMainPlanetRequest
            ? this._multiPlanetary.getPlanetVaultAddress(requestPlanetName)
            : user9cAddress;

        const memo = isMainPlanetRequest ? transactionHash : user9cAddress;

        // Added logging for checking args
        console.log({
          recipient,
          memo,
          amountString,
        });

        const nineChroniclesTxId = await this._ncgTransfer.transfer(
          recipient,
          amountString,
          memo
        );

        await this._monitorStateStore.store("bsc", {
          blockHash,
          txId: transactionHash,
        });
        await this._slackMessageSender.sendMessage(
          new UnwrappedEvent(
            this._explorerUrl,
            this._ncscanUrl,
            this._useNcscan,
            this._etherscanUrl,
            sender,
            user9cAddress,
            amountString,
            nineChroniclesTxId,
            transactionHash,
            isMultiPlanetRequestType,
            requestPlanetName
          )
        );
        await this._exchangeHistoryStore.updateStatus(
          transactionHash,
          TransactionStatus.COMPLETED
        );
        await this._opensearchClient.to_opensearch("info", {
          content: "wNCG -> NCG request success",
          libplanetTxId: nineChroniclesTxId,
          ethereumTxId: transactionHash,
          sender: sender,
          recipient: user9cAddress,
          amount: amount.toNumber(),
          planetName: requestPlanetName,
          network: "BSC",
        });
        await this._opensearchMigrationClient.to_opensearch("info", {
          content: "wNCG -> NCG request success",
          libplanetTxId: nineChroniclesTxId,
          ethereumTxId: transactionHash,
          sender: sender,
          recipient: user9cAddress,
          amount: amount.toNumber(),
          planetName: requestPlanetName,
          network: "BSC",
        });
        console.log("Transferred", nineChroniclesTxId);
      } catch (e) {
        const slackMsgRes = await this._slackMessageSender.sendMessage(
          new UnwrappingFailureEvent(
            this._etherscanUrl,
            sender,
            user9cAddress,
            amountString,
            transactionHash,
            String(e),
            requestPlanetName,
            this._failureSubscribers
          )
        );

        await this._exchangeHistoryStore.updateStatus(
          transactionHash,
          TransactionStatus.FAILED
        );

        await this._spreadsheetClient.to_spreadsheet_burn({
          slackMessageId: `${slackMsgRes?.channel}/p${slackMsgRes?.ts?.replace(
            ".",
            ""
          )}`,
          url: this._etherscanUrl,
          txId: transactionHash,
          sender,
          recipient: String(user9cAddress),
          amount: amountString,
          planetName: requestPlanetName,
          error: String(e),
        });

        await this._opensearchClient.to_opensearch("error", {
          content: "wNCG -> NCG request failure",
          cause: String(e),
          ethereumTxId: transactionHash,
          sender: sender,
          recipient: user9cAddress,
          amount: amount.toNumber(),
          planetName: requestPlanetName,
          network: "BSC",
        });
        await this._opensearchMigrationClient.to_opensearch("error", {
          content: "wNCG -> NCG request failure",
          cause: String(e),
          ethereumTxId: transactionHash,
          sender: sender,
          recipient: user9cAddress,
          amount: amount.toNumber(),
          planetName: requestPlanetName,
          network: "BSC",
        });
        await this._integration.error(
          "Unexpected error during unwrapping NCG",
          {
            errorMessage: String(e),
            sender,
            user9cAddress,
            transactionHash,
            amountString,
          }
        );
      }
    }
  }
}
