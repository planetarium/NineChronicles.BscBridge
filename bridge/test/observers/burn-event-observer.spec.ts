import { EventData } from "web3-eth-contract";
import { INCGTransfer } from "../../src/interfaces/ncg-transfer";
import { IMonitorStateStore } from "../../src/interfaces/monitor-state-store";
import { OpenSearchClient } from "../../src/opensearch-client";
import { TxId } from "../../src/types/txid";
import { BscBurnEventObserver } from "../../src/observers/burn-event-observer";
import { TransactionLocation } from "../../src/types/transaction-location";
import { Integration } from "../../src/integrations";
import { ISlackMessageSender } from "../../src/interfaces/slack-message-sender";
import { SlackMessageSender } from "../../src/slack-message-sender";
import { ISlackChannel } from "../../src/slack-channel";
import { IExchangeHistoryStore } from "../../src/interfaces/exchange-history-store";
import { google } from "googleapis";
import { SpreadsheetClient } from "../../src/spreadsheet-client";
import { FixedExchangeFeeRatioPolicy } from "../../src/policies/exchange-fee-ratio";
import { Decimal } from "decimal.js";
import { MultiPlanetary } from "../../src/multi-planetary";

jest.mock("@slack/web-api", () => {
    return {
        WebClient: jest.fn(() => {
            return {
                chat: {
                    postMessage: jest.fn(),
                },
            };
        }),
    };
});

jest.mock("../../src/opensearch-client", () => {
    return {
        OpenSearchClient: jest.fn(() => {
            return {
                to_opensearch: jest.fn(),
            };
        }),
    };
});

jest.mock("../../src/spreadsheet-client", () => {
    return {
        SpreadsheetClient: jest.fn(() => {
            return {
                to_spreadsheet_mint: jest.fn(),
                to_spreadsheet_burn: jest.fn(),
            };
        }),
    };
});

describe(BscBurnEventObserver.name, () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    const mockNcgTransfer: jest.Mocked<INCGTransfer> = {
        transfer: jest.fn().mockResolvedValue("TX-HASH"),
    };

    const mockSlackChannel: jest.Mocked<ISlackChannel> = {
        sendMessage: jest.fn(),
    };
    const mockSlackMessageSender: ISlackMessageSender = new SlackMessageSender(
        mockSlackChannel
    );

    const mockOpenSearchClient = new OpenSearchClient(
        "https://www.random-url.com",
        "auth",
        "9c-eth-bridge"
    ) as OpenSearchClient & {
        to_opensearch: ReturnType<typeof jest.fn>;
    };

    const mockMonitorStateStore: jest.Mocked<IMonitorStateStore> = {
        load: jest.fn(),
        store: jest.fn(),
    };

    const mockExchangeHistoryStore: jest.Mocked<IExchangeHistoryStore> = {
        put: jest.fn(),
        exist: jest.fn(),
        transferredAmountInLast24Hours: jest.fn(),
    };

    const mockIntegration: jest.Mocked<Integration> = {
        error: jest.fn(),
    };

    const exchangeFeeRatioPolicy = new FixedExchangeFeeRatioPolicy(
        new Decimal(100000),
        new Decimal(50000),
        {
            criterion: new Decimal(1000),
            fee: new Decimal(10),
        },
        {
            range1: new Decimal(0.01),
            range2: new Decimal(0.02),
        }
    );

    const authorize = new google.auth.JWT(
        "randemail@rand.com",
        undefined,
        "rand-key",
        ["spreadsheet-url"]
    );

    const googleSheet = google.sheets({
        version: "v4",
        auth: authorize,
    });

    const mockSpreadSheetClient = new SpreadsheetClient(
        googleSheet,
        "random-id",
        false,
        "slack-url",
        {
            mint: "NCGtoWNCG",
            burn: "WNCGtoNCG",
        },
        exchangeFeeRatioPolicy
    ) as SpreadsheetClient;
    const failureSubscribers = "@gamefi-be";

    const planetIds = {
        odin: "0x100000000000",
        heimdall: "0x100000000001",
    };
    const planetVaultAddress = {
        heimdall: "0xaaaaa6db35d5eff2f0b0758c5ac4c354debaf118",
    };
    const multiPlanetary = new MultiPlanetary(planetIds, planetVaultAddress);

    const observer = new BscBurnEventObserver(
        mockNcgTransfer,
        mockSlackMessageSender,
        mockOpenSearchClient,
        mockSpreadSheetClient,
        mockMonitorStateStore,
        mockExchangeHistoryStore,
        "https://explorer.libplanet.io/9c-internal",
        "https://9cscan.com",
        false,
        "https://ropsten.etherscan.io",
        mockIntegration,
        multiPlanetary,
        failureSubscribers
    );

    describe(BscBurnEventObserver.prototype.notify.name, () => {
        it("should record the block hash even if there is no events", () => {
            observer.notify({
                blockHash: "BLOCK-HASH",
                events: [],
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith(
                "ethereum",
                {
                    blockHash: "BLOCK-HASH",
                    txId: null,
                }
            );
        });

        function makeEvent(
            ncgRecipient: string,
            amount: number,
            txId: TxId
        ): EventData & TransactionLocation {
            return {
                blockHash: "BLOCK-HASH",
                address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                logIndex: 0,
                blockNumber: 0,
                event: "Burn",
                raw: {
                    data: "",
                    topics: [],
                },
                signature: "",
                transactionIndex: 0,
                transactionHash: txId,
                txId,
                returnValues: {
                    _sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    _to: ncgRecipient,
                    amount,
                },
            };
        }
        const ncgRecipient = "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e";

        it("should post slack message every events", async () => {
            const events = [
                makeEvent(ncgRecipient, 1000000000000000000, "TX-A"),
                makeEvent(ncgRecipient, 1200000000000000000, "TX-B"),
                makeEvent(ncgRecipient, 10000000000000000, "TX-C"),
                makeEvent(ncgRecipient, 3225000000000000000, "TX-D"),
            ];

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events,
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith(
                "ethereum",
                {
                    blockHash: "BLOCK-HASH",
                    txId: "TX-D",
                }
            );

            expect(mockExchangeHistoryStore.put.mock.calls).toEqual([
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-A",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 1,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-B",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 1.2,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-C",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 0.01,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-D",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 3.22,
                    },
                ],
            ]);

            expect(mockNcgTransfer.transfer.mock.calls).toEqual([
                [ncgRecipient, "1.00", "TX-A"],
                [ncgRecipient, "1.20", "TX-B"],
                [ncgRecipient, "0.01", "TX-C"],
                [ncgRecipient, "3.22", "TX-D"],
            ]);
        });

        it("should post slack message every events of Multi-Planet Request - Odin", async () => {
            const odinMultiPlanetNcgRecipient = (
                planetIds.odin + ncgRecipient.slice(2)
            ).padEnd(66, "0");
            const events = [
                makeEvent(
                    odinMultiPlanetNcgRecipient,
                    1000000000000000000,
                    "TX-ODIN-A"
                ),
                makeEvent(
                    odinMultiPlanetNcgRecipient,
                    1200000000000000000,
                    "TX-ODIN-B"
                ),
                makeEvent(
                    odinMultiPlanetNcgRecipient,
                    10000000000000000,
                    "TX-ODIN-C"
                ),
                makeEvent(
                    odinMultiPlanetNcgRecipient,
                    3225000000000000000,
                    "TX-ODIN-D"
                ),
            ];

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events,
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith(
                "ethereum",
                {
                    blockHash: "BLOCK-HASH",
                    txId: "TX-ODIN-D",
                }
            );

            expect(mockExchangeHistoryStore.put.mock.calls).toEqual([
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-ODIN-A",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 1,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-ODIN-B",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 1.2,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-ODIN-C",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 0.01,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-ODIN-D",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 3.22,
                    },
                ],
            ]);

            /**
             * If User send wNCG to other planet
             * Recipient: other planet's vault address
             * Memo: user's other planet's 9c Address
             */
            expect(mockNcgTransfer.transfer.mock.calls).toEqual([
                [ncgRecipient, "1.00", "TX-ODIN-A"],
                [ncgRecipient, "1.20", "TX-ODIN-B"],
                [ncgRecipient, "0.01", "TX-ODIN-C"],
                [ncgRecipient, "3.22", "TX-ODIN-D"],
            ]);
        });

        it("should post slack message every events of Heimdall Request", async () => {
            const heimdallMultiPlanetNcgRecipient = (
                planetIds.heimdall + ncgRecipient.slice(2)
            ).padEnd(66, "0");

            const events = [
                makeEvent(
                    heimdallMultiPlanetNcgRecipient,
                    1000000000000000000,
                    "TX-HEIMDALL-A"
                ),
                makeEvent(
                    heimdallMultiPlanetNcgRecipient,
                    1200000000000000000,
                    "TX-HEIMDALL-B"
                ),
                makeEvent(
                    heimdallMultiPlanetNcgRecipient,
                    10000000000000000,
                    "TX-HEIMDALL-C"
                ),
                makeEvent(
                    heimdallMultiPlanetNcgRecipient,
                    3225000000000000000,
                    "TX-HEIMDALL-D"
                ),
            ];

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events,
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith(
                "ethereum",
                {
                    blockHash: "BLOCK-HASH",
                    txId: "TX-HEIMDALL-D",
                }
            );

            expect(mockExchangeHistoryStore.put.mock.calls).toEqual([
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-HEIMDALL-A",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 1,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-HEIMDALL-B",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 1.2,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-HEIMDALL-C",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 0.01,
                    },
                ],
                [
                    {
                        network: "ethereum",
                        tx_id: "TX-HEIMDALL-D",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        recipient: ncgRecipient,
                        timestamp: expect.any(String),
                        amount: 3.22,
                    },
                ],
            ]);

            /**
             * If User send wNCG to other planet
             * Recipient: other planet's vault address
             * Memo: user's other planet's 9c Address
             */
            expect(mockNcgTransfer.transfer.mock.calls).toEqual([
                [planetVaultAddress.heimdall, "1.00", ncgRecipient],
                [planetVaultAddress.heimdall, "1.20", ncgRecipient],
                [planetVaultAddress.heimdall, "0.01", ncgRecipient],
                [planetVaultAddress.heimdall, "3.22", ncgRecipient],
            ]);
        });

        it("slack/opensearch message - snapshot", async () => {
            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        blockHash: "BLOCK-HASH",
                        address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        logIndex: 0,
                        blockNumber: 0,
                        event: "Burn",
                        raw: {
                            data: "",
                            topics: [],
                        },
                        signature: "",
                        transactionIndex: 0,
                        transactionHash: "TX-ID",
                        txId: "TX-ID",
                        returnValues: {
                            _sender:
                                "0x2734048eC2892d111b4fbAB224400847544FC872",
                            _to: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            amount: 1000000000000000000,
                        },
                    },
                ],
            });

            expect(
                mockOpenSearchClient.to_opensearch.mock.calls
            ).toMatchSnapshot();
            expect(mockSlackChannel.sendMessage.mock.calls).toMatchSnapshot();
        });

        it("slack/opensearch 9c transfer error message - snapshot", async () => {
            mockNcgTransfer.transfer.mockImplementationOnce(
                (address, amount, memo) => {
                    throw new Error("mockNcgTransfer.transfer error");
                }
            );

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        blockHash: "BLOCK-HASH",
                        address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        logIndex: 0,
                        blockNumber: 0,
                        event: "Burn",
                        raw: {
                            data: "",
                            topics: [],
                        },
                        signature: "",
                        transactionIndex: 0,
                        transactionHash: "TX-ID",
                        txId: "TX-ID",
                        returnValues: {
                            _sender:
                                "0x2734048eC2892d111b4fbAB224400847544FC872",
                            _to: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            amount: 1000000000000000000,
                        },
                    },
                ],
            });

            expect(
                mockOpenSearchClient.to_opensearch.mock.calls
            ).toMatchSnapshot();
            expect(mockSlackChannel.sendMessage.mock.calls).toMatchSnapshot();
        });

        it("pagerduty 9c transfer error message - snapshot", async () => {
            mockNcgTransfer.transfer.mockImplementationOnce(
                (address, amount, memo) => {
                    throw new Error("mockNcgTransfer.transfer error");
                }
            );

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        blockHash: "BLOCK-HASH",
                        address: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        logIndex: 0,
                        blockNumber: 0,
                        event: "Burn",
                        raw: {
                            data: "",
                            topics: [],
                        },
                        signature: "",
                        transactionIndex: 0,
                        transactionHash: "TX-ID",
                        txId: "TX-ID",
                        returnValues: {
                            _sender:
                                "0x2734048eC2892d111b4fbAB224400847544FC872",
                            _to: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            amount: 1000000000000000000,
                        },
                    },
                ],
            });

            expect(mockIntegration.error.mock.calls).toMatchSnapshot();
        });
    });
});
