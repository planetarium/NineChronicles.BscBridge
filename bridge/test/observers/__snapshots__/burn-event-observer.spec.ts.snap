// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BscBurnEventObserver notify pagerduty 9c transfer error message - snapshot 1`] = `
Array [
  Array [
    "Unexpected error during unwrapping NCG",
    Object {
      "amountString": "1.00",
      "errorMessage": "Error: mockNcgTransfer.transfer error",
      "sender": "0x2734048eC2892d111b4fbAB224400847544FC872",
      "transactionHash": "TX-ID",
      "user9cAddress": "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
    },
  ],
]
`;

exports[`BscBurnEventObserver notify slack/opensearch 9c transfer error message - snapshot 1`] = `
Array [
  Array [
    "error",
    Object {
      "amount": 1,
      "cause": "Error: mockNcgTransfer.transfer error",
      "content": "wNCG -> NCG request failure",
      "ethereumTxId": "TX-ID",
      "network": "BSC",
      "planetName": "odin",
      "recipient": "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
      "sender": "0x2734048eC2892d111b4fbAB224400847544FC872",
    },
  ],
]
`;

exports[`BscBurnEventObserver notify slack/opensearch 9c transfer error message - snapshot 2`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "author_name": "Bridge Error",
          "color": "#ff0033",
          "fallback": "wNCG 0x2734048eC2892d111b4fbAB224400847544FC872 → NCG 0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e failed",
          "fields": Array [
            Object {
              "title": "BSC transaction",
              "value": "https://ropsten.etherscan.io/tx/TX-ID",
            },
            Object {
              "title": "sender (bsc)",
              "value": "0x2734048eC2892d111b4fbAB224400847544FC872",
            },
            Object {
              "title": "recipient (NineChronicles)",
              "value": "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
            },
            Object {
              "title": "amount",
              "value": "1.00",
            },
            Object {
              "title": "Planet Name",
              "value": "odin",
            },
            Object {
              "title": "error",
              "value": "Error: mockNcgTransfer.transfer error",
            },
          ],
        },
      ],
      "text": "[BSC] wNCG → NCG event failed. @gamefi-be",
    },
  ],
]
`;

exports[`BscBurnEventObserver notify slack/opensearch message - snapshot 1`] = `
Array [
  Array [
    "info",
    Object {
      "amount": 1,
      "content": "wNCG -> NCG request success",
      "ethereumTxId": "TX-ID",
      "libplanetTxId": "TX-HASH",
      "network": "BSC",
      "planetName": "odin",
      "recipient": "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
      "sender": "0x2734048eC2892d111b4fbAB224400847544FC872",
    },
  ],
]
`;

exports[`BscBurnEventObserver notify slack/opensearch message - snapshot 2`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "author_name": "Bridge Event",
          "color": "#f5b342",
          "fallback": "wNCG 0x2734048eC2892d111b4fbAB224400847544FC872 → NCG 0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
          "fields": Array [
            Object {
              "title": "9c network transaction",
              "value": "https://explorer.libplanet.io/9c-internal/transaction?TX-HASH",
            },
            Object {
              "title": "Binance Smart Chain network transaction",
              "value": "https://ropsten.etherscan.io/tx/TX-ID",
            },
            Object {
              "title": "sender (bsc)",
              "value": "0x2734048eC2892d111b4fbAB224400847544FC872",
            },
            Object {
              "title": "recipient (NineChronicles)",
              "value": "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
            },
            Object {
              "title": "amount",
              "value": "1.00",
            },
            Object {
              "title": "Planet-Name",
              "value": "odin",
            },
            Object {
              "title": "Network",
              "value": "Binance Smart Chain",
            },
            Object {
              "title": "Not Multi-Planet Request Type",
              "value": "This Transfer Request is not a multi-planet request type.",
            },
          ],
        },
      ],
      "text": "[BSC] wNCG → NCG event occurred.",
    },
  ],
]
`;
